File: mac/Sources/FluentMacKit/Shortcuts.swift
import Foundation

/// The key held for push-to-talk. Values are macOS virtual key codes (Carbon `kVK_*`).
public enum HoldKey: String, CaseIterable, Codable, Sendable {
    case rightOption, fn, rightCommand, rightControl, off

    public var keyCode: UInt16? {
        switch self {
        case .rightOption: 61
        case .fn: 63
        case .rightCommand: 54
        case .rightControl: 62
        case .off: nil
        }
    }

    public var label: String {
        switch self {
        case .rightOption: "Right Option ⌥"
        case .fn: "Fn / Globe"
        case .rightCommand: "Right Command ⌘"
        case .rightControl: "Right Control ⌃"
        case .off: "Off"
        }
    }

    public static func forKeyCode(_ code: UInt16) -> HoldKey? {
        allCases.first { $0.keyCode == code }
    }
}

/// A press-to-start, press-again-to-stop shortcut. Carbon modifier masks, so it can be registered
/// with `RegisterEventHotKey`, which needs no permission at all.
public struct ToggleShortcut: Codable, Equatable, Hashable, Sendable, Identifiable {
    public var keyCode: UInt32
    public var carbonModifiers: UInt32
    public var label: String
    public var id: String { label }

    // Carbon modifier bits.
    public static let cmdKey: UInt32 = 1 << 8
    public static let shiftKey: UInt32 = 1 << 9
    public static let optionKey: UInt32 = 1 << 11
    public static let controlKey: UInt32 = 1 << 12

    public static let space: UInt32 = 49
    public static let d: UInt32 = 2

    public static let presets: [ToggleShortcut] = [
        ToggleShortcut(keyCode: space, carbonModifiers: controlKey | optionKey, label: "⌃⌥ Space"),
        ToggleShortcut(keyCode: space, carbonModifiers: optionKey, label: "⌥ Space"),
        ToggleShortcut(keyCode: d, carbonModifiers: controlKey | optionKey, label: "⌃⌥ D"),
        ToggleShortcut(keyCode: space, carbonModifiers: cmdKey | shiftKey, label: "⇧⌘ Space"),
    ]
    public static let off = ToggleShortcut(keyCode: 0, carbonModifiers: 0, label: "Off")
    public static let `default` = presets[0]

    public var isOff: Bool { self == .off }

    public static func byLabel(_ label: String?) -> ToggleShortcut {
        if label == off.label { return .off }
        return presets.first { $0.label == label } ?? .default
    }
}

/// Push-to-talk timing. A modifier tapped and released quickly, or used together with another
/// key (⌥E for an accent, ⌘C…), is ordinary typing and must never start a dictation.
public struct HoldGesture: Sendable {
    public static let holdDelay: TimeInterval = 0.28

    public enum Event: Sendable { case down(at: TimeInterval), otherKey, up(at: TimeInterval) }
    public enum Action: Equatable, Sendable { case none, armTimer, cancelTimer, stopAndInsert }

    public private(set) var pressedAt: TimeInterval?
    public private(set) var spoiled = false
    public private(set) var dictating = false

    public init() {}

    public mutating func handle(_ e: Event) -> Action {
        switch e {
        case .down(let t):
            pressedAt = t; spoiled = false; dictating = false
            return .armTimer
        case .otherKey:
            guard pressedAt != nil, !dictating else { return .none }
            spoiled = true
            return .cancelTimer
        case .up:
            defer { pressedAt = nil; spoiled = false; dictating = false }
            return dictating ? .stopAndInsert : .cancelTimer
        }
    }

    /// Called when the hold timer fires. True means start recording now.
    public mutating func timerFired() -> Bool {
        guard pressedAt != nil, !spoiled, !dictating else { return false }
        dictating = true
        return true
    }
}

File: src/Fluent.Core/Shortcuts.cs
namespace Fluent.Core;

/// <summary>The key held for push-to-talk (Mac: Right Option / Fn). Right Ctrl is the default: it is
/// never AltGr, and a lone Right Ctrl press does nothing in Windows or in apps.</summary>
public enum HoldKey { RightCtrl, RightAlt, CtrlWin, Off }

public static class HoldKeyInfo
{
    public static string Label(this HoldKey k) => k switch
    {
        HoldKey.RightCtrl => "Right Ctrl",
        HoldKey.RightAlt => "Right Alt",
        HoldKey.CtrlWin => "Ctrl + Win",
        _ => "Off",
    };

    // Virtual-key codes.
    public const int VkLControl = 0xA2, VkRControl = 0xA3, VkRMenu = 0xA5, VkLWin = 0x5B, VkRWin = 0x5C;
    public const int VkEscape = 0x1B;

    /// <summary>Whether the hold key is fully down, given the set of virtual keys currently held.</summary>
    public static bool IsDown(this HoldKey k, IReadOnlySet<int> held) => k switch
    {
        HoldKey.RightCtrl => held.Contains(VkRControl),
        HoldKey.RightAlt => held.Contains(VkRMenu),
        HoldKey.CtrlWin => (held.Contains(VkLControl) || held.Contains(VkRControl)) && (held.Contains(VkLWin) || held.Contains(VkRWin)),
        _ => false,
    };

    /// <summary>Whether <paramref name="vk"/> is one of the keys that make up the hold key.</summary>
    public static bool IsPart(this HoldKey k, int vk) => k switch
    {
        HoldKey.RightCtrl => vk == VkRControl,
        HoldKey.RightAlt => vk == VkRMenu,
        HoldKey.CtrlWin => vk is VkLControl or VkRControl or VkLWin or VkRWin,
        _ => false,
    };
}

/// <summary>A press-to-start, press-again-to-stop shortcut, registered with <c>RegisterHotKey</c>.</summary>
public sealed record ToggleShortcut(uint Modifiers, uint Key, string Label)
{
    // RegisterHotKey modifier bits.
    public const uint Alt = 0x1, Control = 0x2, Shift = 0x4, Win = 0x8, NoRepeat = 0x4000;
    public const uint Space = 0x20, D = 0x44;

    public static readonly IReadOnlyList<ToggleShortcut> Presets =
    [
        new(Control | Alt, Space, "Ctrl + Alt + Space"),
        new(Control | Shift, Space, "Ctrl + Shift + Space"),
        new(Alt, Space, "Alt + Space"),
        new(Control | Alt, D, "Ctrl + Alt + D"),
    ];

    public static readonly ToggleShortcut Off = new(0, 0, "Off");
    public static ToggleShortcut Default => Presets[0];
    public bool IsOff => this == Off;

    public static ToggleShortcut ByLabel(string? label) =>
        label == Off.Label ? Off : Presets.FirstOrDefault(p => p.Label == label) ?? Default;
}

/// <summary>Push-to-talk timing, the Mac's <c>HoldGesture</c> unchanged. A modifier tapped and released
/// quickly, or used together with another key (Ctrl+C…), is ordinary typing and must never start a dictation.</summary>
public sealed class HoldGesture
{
    public static readonly TimeSpan HoldDelay = TimeSpan.FromMilliseconds(280);

    public enum Action { None, ArmTimer, CancelTimer, StopAndInsert }

    public double? PressedAt { get; private set; }
    public bool Spoiled { get; private set; }
    public bool Dictating { get; private set; }

    public Action Down(double at)
    {
        PressedAt = at; Spoiled = false; Dictating = false;
        return Action.ArmTimer;
    }

    public Action OtherKey()
    {
        if (PressedAt is null || Dictating) return Action.None;
        Spoiled = true;
        return Action.CancelTimer;
    }

    public Action Up()
    {
        var result = Dictating ? Action.StopAndInsert : Action.CancelTimer;
        PressedAt = null; Spoiled = false; Dictating = false;
        return result;
    }

    /// <summary>Called when the hold timer fires. True means start recording now.</summary>
    public bool TimerFired()
    {
        if (PressedAt is null || Spoiled || Dictating) return false;
        Dictating = true;
        return true;
    }
}
